File: android/src/main/java/com/appodeal/rnappodeal/RCTAppodealMrecView.java
package com.appodeal.rnappodeal;

import android.app.Activity;
import android.content.res.Resources;
import android.os.Handler;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.appodeal.ads.Appodeal;
import com.appodeal.ads.BannerView;
import com.appodeal.ads.MrecCallbacks;
import com.appodeal.ads.MrecView;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.facebook.react.views.view.ReactViewGroup;


public class RCTAppodealMrecView extends ReactViewGroup implements MrecCallbacks {
    private static final String defaultPlacement = "default";
    private static final Handler handler = new Handler(Looper.getMainLooper());

    private String placement = defaultPlacement;

    @Nullable
    private Runnable showRunnable = null;

    private final Runnable measureAndLayout = () -> {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            child.measure(
                    MeasureSpec.makeMeasureSpec(getMeasuredWidth(), MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(getMeasuredHeight(), MeasureSpec.EXACTLY)
            );
            child.layout(0, 0, child.getMeasuredWidth(), child.getMeasuredHeight());
        }
    };

    public RCTAppodealMrecView(ThemedReactContext context) {
        super(context);
        cacheAdIfNeeded();
    }

    @Override
    public void requestLayout() {
        super.requestLayout();
        post(measureAndLayout);
    }

    public void setPlacement(String placement) {
        this.placement = placement;
    }

    public String getPlacement() {
        return placement;
    }

    public void showBannerView(@NonNull MrecView bannerView) {
        showRunnable = () -> {
            Activity activity = getReactContext().getCurrentActivity();
            if (activity == null) return;

            Resources resources = getReactContext().getResources();
            DisplayMetrics dm = resources.getDisplayMetrics();
            int widthPixels = resources.getDisplayMetrics().widthPixels;
            int heightPixels = dp2px(250, dm);

            LayoutParams bannerLayoutParams = new BannerView.LayoutParams(widthPixels, heightPixels);
            bannerView.setLayoutParams(bannerLayoutParams);
            bannerView.setVisibility(VISIBLE);
            bannerView.setDescendantFocusability(ViewGroup.FOCUS_BLOCK_DESCENDANTS);

            removeAllViews();
            setVisibility(VISIBLE);
            addView(bannerView);

            bannerView.bringToFront();

            String showPlacement = placement == null ? defaultPlacement : placement;
            Appodeal.show(activity, Appodeal.MREC, showPlacement);
        };
        handler.postDelayed(showRunnable, 250L);
    }

    public void hideBannerView(@NonNull MrecView bannerView) {
        removeAllViews();
        ViewGroup parent = (ViewGroup) bannerView.getParent();
        if (parent != null) {
            parent.removeView(bannerView);
        }
        handler.removeCallbacks(showRunnable);
    }

    private void cacheAdIfNeeded() {
        if (!Appodeal.isAutoCacheEnabled(Appodeal.MREC)) {
            Activity activity = getReactContext().getCurrentActivity();
            if (activity != null) {
                Appodeal.cache(activity, Appodeal.MREC);
            }
        }
    }

    private ReactContext getReactContext() {
        return (ReactContext) getContext();
    }

    private RCTEventEmitter getEmitter() {
        return getReactContext().getJSModule(RCTEventEmitter.class);
    }

    private int dp2px(int dp, DisplayMetrics dm) {
        return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, dm));
    }

    @Override
    public void onMrecLoaded(boolean isPrecache) {
        WritableMap params = Arguments.createMap();
        params.putInt("height", 250);
        params.putBoolean("isPrecache", isPrecache);
        getEmitter().receiveEvent(getId(), "onMrecLoaded", params);
    }

    @Override
    public void onMrecFailedToLoad() {
        getEmitter().receiveEvent(getId(), "onMrecFailedToLoad", null);
    }

    @Override
    public void onMrecClicked() {
        getEmitter().receiveEvent(getId(), "onMrecClicked", null);
    }

    @Override
    public void onMrecExpired() {
        getEmitter().receiveEvent(getId(), "onMrecExpired", null);
    }

    @Override
    public void onMrecShowFailed() { }

    @Override
    public void onMrecShown() { }
}

File: android/src/main/java/com/appodeal/rnappodeal/RCTAppodealMrecView.kt
package com.appodeal.rnappodeal

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.DisplayMetrics
import android.util.Log
import android.util.TypedValue
import android.view.ViewGroup
import android.widget.FrameLayout
import com.appodeal.ads.Appodeal
import com.appodeal.ads.MrecView
import com.appodeal.rnappodeal.callbacks.RNAppodealEventHandler
import com.appodeal.rnappodeal.constants.MrecEvents
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.events.Event
import com.facebook.react.uimanager.events.EventDispatcher
import com.facebook.react.views.view.ReactViewGroup
import kotlin.math.roundToInt

class RCTAppodealMrecView(context: Context) : ReactViewGroup(context), RNAppodealEventHandler {

    private val surfaceId: Int by lazy { UIManagerHelper.getSurfaceId(reactContext) }
    private val uiHandler: Handler by lazy { Handler(Looper.getMainLooper()) }

    var placement: String = "default"
        set(value) {
            field = value
            cacheAdIfNeeded()
        }

    private var showRunnable: Runnable? = null

    private val measureAndLayout = Runnable {
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            child.measure(
                MeasureSpec.makeMeasureSpec(measuredWidth, MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(measuredHeight, MeasureSpec.EXACTLY)
            )
            child.layout(0, 0, child.measuredWidth, child.measuredHeight)
        }
    }

    override fun requestLayout() {
        super.requestLayout()
        post(measureAndLayout)
    }

    fun showMrecView(bannerView: MrecView) {
        val runnable = Runnable {
            val activity = reactContext.currentActivity ?: return@Runnable

            // MREC is always 300x250
            val widthPx = dp2px(300, reactContext.resources.displayMetrics)
            val heightPx = dp2px(250, reactContext.resources.displayMetrics)

            bannerView.layoutParams = FrameLayout.LayoutParams(widthPx, heightPx)
            bannerView.visibility = VISIBLE
            bannerView.descendantFocusability = FOCUS_BLOCK_DESCENDANTS

            removeAllViews()
            this.visibility = VISIBLE
            addView(bannerView)
            bannerView.bringToFront()

            try {
                // Ensure this mrec view is brought to front in parent hierarchy
                (parent as? ViewGroup)?.bringChildToFront(this)
            } catch (e: Exception) {
                Log.e("RTCAppodealMrec", "Error bringing banner view to front", e)
            }

            Appodeal.show(activity, Appodeal.MREC, placement)
        }.also { showRunnable = it }
        uiHandler.postDelayed(runnable, 250L)
    }

    fun hideMrecView(bannerView: MrecView) {
        removeAllViews()
        (bannerView.parent as? ViewGroup)?.removeView(bannerView)
        showRunnable?.let { uiHandler.removeCallbacks(it) }
    }

    private fun cacheAdIfNeeded() {
        if (!Appodeal.isAutoCacheEnabled(Appodeal.MREC)) {
            reactContext.currentActivity?.let { activity ->
                Appodeal.cache(activity, Appodeal.MREC)
            }
        }
    }

    override fun handleEvent(event: String, params: WritableMap?) {
        when (event) {
            MrecEvents.ON_MREC_LOADED -> dispatchFabricEvent(id, MrecEvents.ON_AD_LOADED, params)
            MrecEvents.ON_MREC_FAILED_TO_LOAD -> dispatchFabricEvent(id, MrecEvents.ON_AD_FAILED_TO_LOAD, params)
            MrecEvents.ON_MREC_CLICKED -> dispatchFabricEvent(id, MrecEvents.ON_AD_CLICKED, params)
            MrecEvents.ON_MREC_EXPIRED -> dispatchFabricEvent(id, MrecEvents.ON_AD_EXPIRED, params)
            else -> Unit
        }
    }

    private fun dispatchFabricEvent(
        viewId: Int,
        eventName: String,
        params: WritableMap?
    ) {
        val dispatcher: EventDispatcher? =
            UIManagerHelper.getEventDispatcherForReactTag(reactContext, viewId)
        dispatcher?.dispatchEvent(OnViewEvent(surfaceId, viewId, eventName, params))
    }

    private class OnViewEvent(
        surfaceId: Int,
        viewId: Int,
        private val eventName: String,
        private val payload: WritableMap?
    ) : Event<OnViewEvent>(surfaceId, viewId) {
        override fun getEventName(): String = eventName
        override fun getEventData(): WritableMap? = payload
    }

    private val reactContext: ReactContext
        get() = context as ReactContext
}

private fun dp2px(dp: Int, dm: DisplayMetrics): Int =
    TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), dm).roundToInt() 